lib/msf/core/opt_rhost_url.rb

```diff
@@ -0,0 +1,75 @@
+# -*- coding: binary -*-
+
+module Msf
+  ###
+  #
+  # RHOST URL option.
+  #
+  ###
+  class OptRhostUrl < OptBase
+    def type
+      'rhost url'
+    end
+
+
+    def normalize(value)
+      return unless value
+
+      uri = get_uri(value)
+      option_hash = {}
+      # Blank this out since we don't know if this new value will have a `VHOST` to ensure we remove the old value
+      option_hash['VHOST'] = nil
+
+      option_hash['RHOSTS'] = uri.hostname
+      option_hash['RPORT'] = uri.port
+      option_hash['SSL'] = %w{ssl https}.include?(uri.scheme)
+
+      # Both `TARGETURI` and `URI` are used as datastore options to denote the path on a uri
+      option_hash['TARGETURI'] = uri.path || '/'
+      option_hash['URI'] = uri.path || '/'
+
+      if uri.scheme and %{http https}.include?(uri.scheme)
+        option_hash['VHOST'] = uri.hostname unless Rex::Socket.is_ip_addr?(uri.hostname)
+        option_hash['HttpUsername'] = uri.user.to_s
+        option_hash['HttpPassword'] = uri.password.to_s
+      end
+
+      option_hash
+    end
+
+    def valid?(value, check_empty: false)
+      return true unless value
+      uri = get_uri(value)
+      false unless uri.host != nil and uri.port != nil
```

Comment:
```suggestion
      return false if uri.host.nil? && uri.port.nil?
```

Did you mean to `return`? What happens with the `super`?